Procedure additions/mutations => Variable updates

AnmAtAnm opened the pull request:
* Refactored variable storage in the WorkspaceStats to use a new VariableInfo object.
 * Refactored procedure mutators to use a shared ProcedureInfo object for the mutator description.
 * Procedure blocks additions and changes now automatically create variables. Both of these produce changes on the ProcedureCustomCategory and VariableCustomCategory, respectively.
 * Fail gracefully on autoload.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-android/615)
<!-- Reviewable:end -->

RoboErikG commented:
Review status: 0 of 22 files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 322 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8MB01iJY5lKA0at0P:-Km8MB01iJY5lKA0at0Q:b-f20ydv) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L322)):*
> ```Java
>         final ProcedureInfo oldProcInfo = definitionMutator.getProcedureInfo();
>         final String newProcedureName = updatedProcedureInfo.getProcedureName();
>         final boolean isFuncRename = originalProcedureName.equals(newProcedureName);
> ```

This needs a !, it's a rename if the names are not equal.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 355 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8Ncq0zC2FU-ECd7Pu:-Km8Ncq1KXWDWYofOUZf:b-qa4o6l) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L355)):*
> ```Java
> 
>                 // Mutate each procedure call
>                 List<Block> procedureCalls = mProcedureReferences.get(originalProcedureName);
> ```

in a rename the old procedure name is never removed from mProcedureReferences or mProcedureNameManager. You may want a variation on removeDefinition that takes the procedure name and call it at the end to do cleanup of the old name.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 96 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8Q9phdTqNSkhT3JZt:-Km8Q9phdTqNSkhT3JZu:b-onee48) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L96)):*
> ```Java
>                     VariableInfoImpl varInfo = mVariableInfoMap.get(i);
>                     varInfo.removeProcedure(oldName);
>                     varInfo.addProcedure(newName);
> ```

Is this adding the procedure to every variable  in the workspace? I think you want to remove it from all the variables in the old list and then add it to all the variables in the new list, regardless of the name changing.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 257 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8RbHaAaIy9e8-EbaS:-Km8RbHaAaIy9e8-EbaT:b2z5h8s) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L257)):*
> ```Java
>      * Attempts to add a variable to the workspace.
>      * @param requestedName The preferred variable name. Usually the user name.
>      * @param allowRename Whether the variable name should be rename
> ```

"can be renamed"

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 276 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8SwLX84fLUCUH1tHc:-Km8SwLYSnEh-sPNZdi5:bneg9mp) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L276)):*
> ```Java
>         final String mDisplayName;
>         /** FieldVariables that are set to the variable. */
>         ArrayList<WeakReference<FieldVariable>> mFields = null;
> ```

Why do you keep creating and deleting mFields and mProcedures? Why not just leave them at 0 when there's nothing in them?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 326 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8SKzo1DI0bD19B1zD:-Km8SKzo1DI0bD19B1zE:b-cewayh) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L326)):*
> ```Java
>         }
> 
>         boolean removeField(FieldVariable newField) {
> ```

s/newField/oldField

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 340 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8SS46ltBA4a-S2rwX:-Km8SS46ltBA4a-S2rwY:b-tgb4u1) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L340)):*
> ```Java
>                 if (field == newField) {
>                     mFields.remove(i);
>                     --count;
> ```

nit: --count is unnecessary since you're returning. Fine to leave in for future safety. Missing check for mFields.isEmpty if this is the last variable in the list.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/ProcedureInfo.java, line 53 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8UUXY_4gGvDEixS-V:-Km8UUXZMb4nFs8kE5g1:b-n0jdrs) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/model/ProcedureInfo.java#L53)):*
> ```Java
>         return mArguments;
>     }
>     
> ```

nit ws

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java, line 262 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8X1IVQb1IntxTUQpG:-Km8X1IVQb1IntxTUQpH:b-qotv2o) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java#L262)):*
> ```Java
>         List<String> arguments = mProcedureInfo.getArguments();
>         if (!arguments.isEmpty()) {
>             sb.append("with:"); // message BKY_PROCEDURES_BEFORE_PARAMS
> ```

nit, replace 'with:' with a string resource.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/615)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Just some questions/nits then <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 0 of 23 files reviewed at latest revision, 12 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 360 at r2](https://reviewable.io:443/reviews/google/blockly-android/615#-Km97JdJED0oyvlXsfhP:-Km97JdJED0oyvlXsfhQ:b-gfmf74) ([raw file](https://github.com/google/blockly-android/blob/c50d24e5a2fcf2b9d20989f9e2a329132250d1f4/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L360)):*
> ```Java
>                 definitionMutator.mutate(updatedProcedureInfo);
>                 if (isFuncRename) {
>                     mProcedureNameManager.remove(originalProcedureName);
> ```

Fine for now, but might want a TODO to call a cleanup method at the end in case more things get added later.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 276 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8SwLX84fLUCUH1tHc:-Km98q-BABFOvXeoT54s:b-u2poym) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L276)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Why do you keep creating and deleting mFields and mProcedures? Why not just leave them at 0 when there's nothing in them?
</blockquote></details>

Still needs followup.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 90 at r2](https://reviewable.io:443/reviews/google/blockly-android/615#-Km97urDklMej6UBm819:-Km97urDklMej6UBm81A:b-g46bhr) ([raw file](https://github.com/google/blockly-android/blob/c50d24e5a2fcf2b9d20989f9e2a329132250d1f4/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L90)):*
> ```Java
>         public void onProcedureMutated(ProcedureInfo oldInfo, ProcedureInfo newInfo) {
>             String oldName = oldInfo.getProcedureName();
>             String newName = newInfo.getProcedureName();
> ```

Just to check, does this still work if the list of args is what changed?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 258 at r2](https://reviewable.io:443/reviews/google/blockly-android/615#-Km98POIZb0LvYjJNP7r:-Km98POIZb0LvYjJNP7s:b936c0p) ([raw file](https://github.com/google/blockly-android/blob/c50d24e5a2fcf2b9d20989f9e2a329132250d1f4/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L258)):*
> ```Java
>      * Attempts to add a variable to the workspace.
>      * @param requestedName The preferred variable name. Usually the user name.
>      * @param allowRename Whether the variable name can be rename.
> ```

nit: missing d on renamed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/615#-:-Km98yso2_T9T7ix9RpN:b-1fqnzq)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 0 of 23 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 322 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8MB01iJY5lKA0at0P:-Km8ayk7If7TEtOiJRwk:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L322)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

This needs a !, it's a rename if the names are not equal.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 355 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8Ncq0zC2FU-ECd7Pu:-Km8yrtbwnurzzkX7q60:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L355)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

in a rename the old procedure name is never removed from mProcedureReferences or mProcedureNameManager. You may want a variation on removeDefinition that takes the procedure name and call it at the end to do cleanup of the old name.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 96 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8Q9phdTqNSkhT3JZt:-Km8ytmz5qIM0sHb9-A4:bt2vr1b) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L96)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Is this adding the procedure to every variable  in the workspace? I think you want to remove it from all the variables in the old list and then add it to all the variables in the new list, regardless of the name changing.
</blockquote></details>

Fixed.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 257 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8RbHaAaIy9e8-EbaS:-Km8z2UWATwp70chol2l:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L257)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

"can be renamed"
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 276 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8SwLX84fLUCUH1tHc:-Km8z4gl8etDjttcFraC:bwmtpfb) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L276)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Why do you keep creating and deleting mFields and mProcedures? Why not just leave them at 0 when there's nothing in them?
</blockquote></details>

Memory usage.  While the lists are tiny, the VariableInfo objects are long lived objects where many/most variable probably don't need both lists, and if they list at least 1, I would not expect them to go down to zero (or vice versa) very often.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 326 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8SKzo1DI0bD19B1zD:-Km9-1Q2r-BfPnTSmZg-:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L326)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

s/newField/oldField
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 340 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8SS46ltBA4a-S2rwX:-Km9-8QzG_lxGurpZMBN:bss4a2b) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L340)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nit: --count is unnecessary since you're returning. Fine to leave in for future safety. Missing check for mFields.isEmpty if this is the last variable in the list.
</blockquote></details>

Removed.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/ProcedureInfo.java, line 53 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8UUXY_4gGvDEixS-V:-Km9-htqJWQUZoye09B5:b-udog60) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/model/ProcedureInfo.java#L53)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nit ws
</blockquote></details>

Weird.  My editor was configured to remove these.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java, line 262 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8X1IVQb1IntxTUQpG:-Km91G8jJSoKLKo-j3W8:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java#L262)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nit, replace 'with:' with a string resource.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/615)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 0 of 25 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 100 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE13K8xefgrjpxsyGf:-KmE13K8xefgrjpxsyGg:b-qodg5c) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L100)):*
> ```Java
>     private final NameManager mProcedureNameManager = new NameManager.ProcedureNameManager();
> 
>     // Used determine the visibility of procedures_ifreturn block in the ProcedureCustomCategory.
> ```

nit, missing 'to' after 'Used'

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 165 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE10teLgEx9_RMlxmj:-KmE10tfDTjMJgfnaBlt:bdvek9b) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L165)):*
> ```Java
> 
>     /**
>      * @return All the registered procedure definition blocks, keyed by
> ```

keyed by...?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 257 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8RbHaAaIy9e8-EbaS:-KmE1v35TtBMONR6ko_M:bgaftia) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L257)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still not done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 90 at r2](https://reviewable.io:443/reviews/google/blockly-android/615#-Km97urDklMej6UBm819:-KmE1gmC-65RARqA4bB-:b-3emgn7) ([raw file](https://github.com/google/blockly-android/blob/c50d24e5a2fcf2b9d20989f9e2a329132250d1f4/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L90)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Just to check, does this still work if the list of args is what changed?
</blockquote></details>

Still needs followup. Does this work correctly if the name doesn't change but the number or names of parameters does?

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureCallMutator.java, line 147 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE2rZC6IYJ8TB0erRk:-KmE2rZC6IYJ8TB0erRl:bk97307) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureCallMutator.java#L147)):*
> ```Java
>      * Updates the block using the mutation in the XML.
>      *
>      * @param parser The parser with the {@code <mutation>} element.
> ```

Do you need to escape the lt/gt signs?

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java, line 157 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE3B95KebmkH9cV387:-KmE3B95KebmkH9cV388:b-tfylmc) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java#L157)):*
> ```Java
>     /**
>      * Called when the mutator is attached to a block. It will make sure the procedure name on the
>      * block's name field is in synch with the mutator's PRocedureInfo, and register a listener on
> ```

nits, usually sync without the h, PRocedureInfo capitlization typo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/615)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 0 of 25 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 360 at r2](https://reviewable.io:443/reviews/google/blockly-android/615#-Km97JdJED0oyvlXsfhP:-Km99bPTexsUFBbAbXjT:b-h58sr0) ([raw file](https://github.com/google/blockly-android/blob/c50d24e5a2fcf2b9d20989f9e2a329132250d1f4/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L360)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Fine for now, but might want a TODO to call a cleanup method at the end in case more things get added later.
</blockquote></details>

I'm not clear on the case you're considering.  I'll ask tomorrow.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 100 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE13K8xefgrjpxsyGf:-KmEFL2g4JH1tqXsXhiC:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L100)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nit, missing 'to' after 'Used'
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 165 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE10teLgEx9_RMlxmj:-KmEFXoQC1jBbfLVGHXo:bmxa49t) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L165)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

keyed by...?
</blockquote></details>

...canonical procedure name.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 257 at r1](https://reviewable.io:443/reviews/google/blockly-android/615#-Km8RbHaAaIy9e8-EbaS:-KmEEPVbkC2I6y4l0_5O:b-e870u5) ([raw file](https://github.com/google/blockly-android/blob/921ffc910212fa0efc7c669d31335491c55a2c64/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L257)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Still not done.
</blockquote></details>

Done for realzies.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 90 at r2](https://reviewable.io:443/reviews/google/blockly-android/615#-Km97urDklMej6UBm819:-KmE0G8yp8h0mQJpi6HO:b9fwxiu) ([raw file](https://github.com/google/blockly-android/blob/c50d24e5a2fcf2b9d20989f9e2a329132250d1f4/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L90)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Still needs followup. Does this work correctly if the name doesn't change but the number or names of parameters does?
</blockquote></details>

TODO per offline conversation.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java, line 258 at r2](https://reviewable.io:443/reviews/google/blockly-android/615#-Km98POIZb0LvYjJNP7r:-KmEENxWGZJ0S0M6NFVA:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c50d24e5a2fcf2b9d20989f9e2a329132250d1f4/blocklylib-core/src/main/java/com/google/blockly/android/control/WorkspaceStats.java#L258)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nit: missing d on renamed.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureCallMutator.java, line 147 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE2rZC6IYJ8TB0erRk:-KmEEtKcMA4weAVn6-lq:bt2luyq) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureCallMutator.java#L147)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Do you need to escape the lt/gt signs?
</blockquote></details>

Inside a {@code} it should be fine.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java, line 157 at r3](https://reviewable.io:443/reviews/google/blockly-android/615#-KmE3B95KebmkH9cV387:-KmEFBljYU8RVLTaYXZL:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/48dae29762f2bfd709434910f2952365e3b90461/blocklylib-core/src/main/java/com/google/blockly/model/mutator/ProcedureDefinitionMutator.java#L157)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nits, usually sync without the h, PRocedureInfo capitlization typo.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/615)*
<!-- Sent from Reviewable.io -->
